klib/userdata.c

```diff
@@ -0,0 +1,707 @@
+static boolean metadata_server_reachable(void);
+
+/* Parse a line for VARNAME=value format and set environment variable */
+static void parse_and_set_env_line(buffer line)
+{
+    if (buffer_length(line) == 0)
+        return;
+
+    /* Skip comments and blank lines */
+    char first = byte(line, 0);
+    if (first == '#' || first == '\n' || first == '\r')
+        return;
+
+    int eq_pos = buffer_strchr(line, '=');
+    if (eq_pos <= 0)
+        return;  /* no equals sign or it's at the start */
+
+    /* Extract variable name */
+    buffer name_buf = allocate_buffer(userdata_heap, eq_pos);
+    if (name_buf == INVALID_ADDRESS)
+        return;
+
+    buffer_write(name_buf, buffer_ref(line, 0), eq_pos);
+
+    /* Extract value (rest of line after '=') */
+    int value_len = buffer_length(line) - eq_pos - 1;
+
+    /* Trim trailing newline/carriage return */
+    while (value_len > 0) {
+        char last = byte(line, eq_pos + value_len);
+        if (last == '\n' || last == '\r')
+            value_len--;
+        else
+            break;
+    }
+
+    if (value_len > 0) {
+        buffer value_buf = allocate_buffer(userdata_heap, value_len);
+        if (value_buf != INVALID_ADDRESS) {
+            buffer_write(value_buf, buffer_ref(line, eq_pos + 1), value_len);
+
+            /* Set environment variable */
+            symbol var_sym = intern(name_buf);
+            set(get_environment(), var_sym, value_buf);
+
+            userdata_debug("Set environment variable: %b=%b", name_buf, value_buf);
+        }
+    }
+
+    deallocate_buffer(name_buf);
+}
+
+/* Parse userdata content for environment variables */
+static void parse_userdata(buffer content)
+{
+    if (!content || buffer_length(content) == 0)
+        return;
+
+    buffer line = allocate_buffer(userdata_heap, 512);
+    if (line == INVALID_ADDRESS)
+        return;
+
+    bytes offset = 0;
+    bytes content_len = buffer_length(content);
+
+    while (offset < content_len) {
+        buffer_clear(line);
+
+        /* Read until newline */
+        while (offset < content_len) {
+            char c = byte(content, offset++);
+            push_u8(line, c);
+            if (c == '\n')
+                break;
+        }
+
+        parse_and_set_env_line(line);
+    }
+
+    deallocate_buffer(line);
+}
+
+/* Decode base64 content (e.g. for Azure) */
+static buffer decode_base64(buffer encoded)
+{
+    if (!encoded || buffer_length(encoded) == 0)
+        return 0;
+
+    /* Calculate required output size */
+    bytes encoded_len = buffer_length(encoded);
+    bytes decoded_len = (encoded_len * 3) / 4 + 4;  /* upper bound */
+
+    buffer decoded = allocate_buffer(userdata_heap, decoded_len);
+    if (decoded == INVALID_ADDRESS)
+        return 0;
+
+    size_t actual_len;
+    int ret = mbedtls_base64_decode(buffer_ref(decoded, 0), decoded_len,
+                                     &actual_len,
+                                     buffer_ref(encoded, 0), encoded_len);
```

Comment:
The indentation should follow the opening parenthesis.
By the way, we use only space characters (no tabs) for indentation.